rfcs/0049-flakes.md

```diff
@@ -0,0 +1,709 @@
+  * `file:///home/my-user/some-repo/some-repo.git`
+
+* Local paths. These have the form
+
+      <path>(\?<params)?
+
+  where `<path>` must refer to (a subdirectory of) a Git repository.
+  These differ from `file://` flake references in a few ways:
+
+  * They refer to the working tree (unless an explicit `rev` or `ref`
+    is specified), so evaluation can access dirty files. (Dirty files
+    are files that are tracked by git but have uncommitted changes.)
+
+  * The `dir` parameter is automatically derived. For example, if
+    `/foo/bar` is a Git repository, then the flake reference
+    `/foo/bar/flake` is equivalent to `/foo/bar?dir=flake`.
+
+* GitHub repositories. These are downloaded as tarball archives,
+  rather than through Git. This is often much faster and uses less
+  disk space since it doesn't require fetching the entire history of
+  the repository. On the other hand, it doesn't allow incremental
+  fetching (but full downloads are often faster than incremental
+  fetches!). The syntax is:
+
+      github:<owner>/<repo>(/<rev-or-ref>)?(\?<params>)?
+
+  `<rev-or-ref>` specifies the name of a branch or tag (`ref`), or a
+  commit hash (`rev`). Note that unlike GitHub, Git allows fetching by
+  commit hash without specifying a branch or tag.
+
+  The only supported parameter is `dir` (see above).
+
+  Some examples:
+
+  * `github:edolstra/dwarffs`
+  * `github:edolstra/dwarffs/unstable`
+  * `github:edolstra/dwarffs/d3f2baba8f425779026c6ec04021b2e927f61e31`
+
+* Indirections through the flake registry. These have the form
+
+      <flake-id>(/<rev-or-ref>(/rev)?)?
+
+  These perform a lookup of `<flake-id>` in the flake registry. The
+  specified `rev` and/or `ref` are then merged with the entry in the
+  registry. (See below.) For example, `nixpkgs` and
+  `nixpkgs/release-19.03` are indirect flake references.
+
+In the future, we should also add tarball flake references
+(e.g. `https://example.org/nixpkgs.tar.xz`). It would also be
+straight-forward to add flake references to Mercurial repositories
+since Nix already has Mercurial support.
+
+## Flake registries
+
+Flake registries map symbolic flake identifiers (e.g. `nixpkgs`) to
+"direct" flake references (i.e. any type of flake reference that's not
+an indirection). This is a convenience to users, allowing them to do
+
+    nix run nixpkgs:hello
+
+rather than
+
+    nix run github:NixOS/nixpkgs:hello
+
+There are multiple registries:
+
+* The global registry
+  https://raw.githubusercontent.com/NixOS/flake-registry/master/flake-registry.json. (This
+  location can be overriden via the `flake-registry` option.) Nix
+  automatically fetches this registry periodically. The check interval
+  is determined by the `tarball-ttl` option.
+
+* The local registry `~/.config/nix/registry.json`.
+
+* Registry entries specified on the command line.
+
+A registry is a JSON file that looks like this:
```

Comment:
Why not Nix? Using JSON and a management CLI rather than encouraging editing the file for this feels a bit at odds with how Nix does everything else.

Reply:
Nix is turing complete, disallowing (recursive) functions has a lot of merits to disallow whole classes of problems.

Reply:
It's more that Nix code is hard to patch with new values if there are functions in there. If it was a subset of Nix with only the Object Notation then it would be easy.

Reply:
However that still means we’d need nix-specific tools to do any kind of automated housekeeping, while there’s already plenty of those for JSON

Reply:
If we want automated housekeeping, I’m all for JSON. The question I’m asking is “do we?”.

Reply:
To turn it around: what's the argument for making it a Nix file?

Reply:
> Nix is turing complete, disallowing (recursive) functions has a lot of merits to disallow whole classes of problems.
True, but this affects a large part of the nix ecosystem beyond just a registry.
>Nix code is hard to patch with new values
I don't believe there exists anywhere else where Nix will overwrite a nix file that you have written. Nix shouldn't be changing your Nix files.
>what's the argument for making it a Nix file
The configuration for the entire rest of the system, bar .nix-channels, is configured using Nix. It's just weird that the channels, or this flake system is using an entirely separate configuration from the rest of the system. This makes those systems not feel first class as you have to configure them separately. Personally, I would rather see these systems getting added to the actual configuration files for the system (/etc/nixos/configuration.nix) or users (like home-manager's home.nix). Having the configuration split up into it's own little system as a .nix-channels or a registry.json just complicates reproducing machines. It feels like it's a mutable part to the system being able to change your channels which can cause the same nix file create two different systems based off what channel and what version of it you are on. I will admit the drawback to doing it this way is that it is harder to upgrade as you'll need to update the branches and which revision of them you want your system to use. Perhaps it may be worth making a trade off between reproducability and convenience by allowing the version of the channel you want your system to be on to be optional (yes I know this is similar to the idea of pinning).

Reply:
The upside of using Nix is that registries can be override/expanded. The downside is that it's harder to generate them. Should we generate registries? I think we'd only need their output, which can be JSON.

Reply:
I think for the flake MVP we should stick to JSON. We can always introduce `.nix` registries in the future if there is demand, but since they're simple `identifier -> flake location` mappings, I don't think it will be needed.